crypto/pkcs7/bio/md.c

```diff
@@ -0,0 +1,178 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 OR ISC
+
+#include <errno.h>
+#include <openssl/buffer.h>
+#include <openssl/evp.h>
+#include <openssl/pkcs7.h>
+#include "../../internal.h"
+#include "../crypto/bio/internal.h"
+#include "../internal.h"
+
+// |md_write| and |md_read| both contribute to the digest, |md_gets| returns
+// digest contents
+static int md_write(BIO *h, char const *buf, int num);
+static int md_read(BIO *h, char *buf, int size);
+static int md_gets(BIO *h, char *str, int size);
+static long md_ctrl(BIO *h, int cmd, long arg1, void *arg2);
+static int md_new(BIO *h);
+static int md_free(BIO *data);
+
+
+static const BIO_METHOD methods_md = {
+    BIO_TYPE_MD, "message digest", md_write, md_read, /*puts*/ NULL,
+    md_gets,     md_ctrl,          md_new,   md_free, /*callback_ctrl*/ NULL,
+};
+const BIO_METHOD *BIO_f_md(void) { return &methods_md; }
+
+static int md_new(BIO *b) {
+  GUARD_PTR(b);
+  EVP_MD_CTX *ctx;
+
+  ctx = EVP_MD_CTX_new();
+  if (ctx == NULL) {
+    return 0;
+  }
+
+  BIO_set_data(b, ctx);
+
+  return 1;
+}
+
+static int md_free(BIO *b) {
+  GUARD_PTR(b);
+  EVP_MD_CTX_free(BIO_get_data(b));
+  BIO_set_data(b, NULL);
+  BIO_set_init(b, 0);
+
+  return 1;
+}
+
+static int md_read(BIO *b, char *out, int outl) {
+  GUARD_PTR(b);
+  GUARD_PTR(out);
+  int ret = 0;
+  EVP_MD_CTX *ctx;
+  BIO *next;
+
+  ctx = BIO_get_data(b);
+  next = BIO_next(b);
+
+  if ((ctx == NULL) || (next == NULL) || outl <= 0) {
+    return 0;
+  }
+
+  ret = BIO_read(next, out, outl);
+  if (BIO_get_init(b)) {
+    if (ret > 0) {
+      if (EVP_DigestUpdate(ctx, (unsigned char *)out, ret) <= 0) {
+        return -1;
```

Comment:
If this ever returns an error, should we somehow mark this `BIO` and being not OK for subsequent use? In this case, `next` has provided new data that can not be reprocessed on any subsequent call.

Reply:
OpenSSL [doesn't do this](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1-stable/crypto/evp/bio_md.c#L98), but i agree philosophically that we should fail loudly and not silently corrupt the digest.

`EVP_DigestUpdate` 's documentation [stipulates](https://github.com/aws/aws-lc/blob/cf969b57ae1fcef117bbbe32a8a43f53389f71a8/include/openssl/digest.h#L165) an interface contract of "It returns one". Looking at [the implementation](https://github.com/aws/aws-lc/blob/cf969b57ae1fcef117bbbe32a8a43f53389f71a8/crypto/fipsmodule/digest/digest.c#L272), the only case where it can break that contract is if an `update` function pointer isn't configured on `ctx`. I believe we can trust that to [always be populated](https://github.com/aws/aws-lc/blob/cf969b57ae1fcef117bbbe32a8a43f53389f71a8/crypto/fipsmodule/digest/digest.c#L234) as long as `ctx` has been initialized by `EVP_DigestInit_ex`.

So, in other words, it looks like `EVP_DigestUpdate` can only fail if `ctx` hasn't been initialized. This makes me consider... should we fail `b`'s reads/writes before `next`'s IO if `b` hasn't been initialized? this would address your concern here and make it harder to misuse the `BIO_f_md`.